pkg/controller/externalnode/controller.go

```diff
@@ -0,0 +1,384 @@
+		return nil
+	}
+	ee := genExternalEntity(eeName, en)
+	err := c.addExternalEntity(ee)
+	if err != nil {
+		return err
+	}
+	c.syncedExternalNode.Add(en)
+	return nil
+}
+
+func (c *ExternalNodeController) addExternalEntity(ee *v1alpha2.ExternalEntity) error {
+	_, err := c.crdClient.CrdV1alpha2().ExternalEntities(ee.Namespace).Create(context.TODO(), ee, metav1.CreateOptions{})
+	if errors.IsAlreadyExists(err) {
+		klog.Warningf("Updating ExternalEntity namespace %s, name %s as it already exists", ee.Namespace, ee.Name)
+		return c.updateExternalEntity(ee)
+	}
+	return err
+}
+
+func (c *ExternalNodeController) updateExternalNode(preEn *v1alpha1.ExternalNode, curEn *v1alpha1.ExternalNode) error {
+	if reflect.DeepEqual(preEn.Spec.Interfaces, curEn.Spec.Interfaces) && reflect.DeepEqual(preEn.Labels, curEn.Labels) {
+		return nil
+	}
+	// Delete the previous ExternalEntity and create a new one if the name of the generated ExternalEntity is changed.
+	// Otherwise, update the ExternalEntity.
+	if preEn.Spec.Interfaces[0].Name != curEn.Spec.Interfaces[0].Name {
+		preEEName := genExternalEntityName(preEn)
+		if preEEName != "" {
+			err := c.deleteExternalEntity(preEn.Namespace, preEEName)
+			if err != nil {
+				return err
+			}
+		}
+		curEEName := genExternalEntityName(curEn)
+		if curEEName != "" {
+			curEE := genExternalEntity(curEEName, curEn)
+			err := c.addExternalEntity(curEE)
+			if err != nil {
+				return err
+			}
+		}
+	} else {
+		preIPs := sets.NewString(preEn.Spec.Interfaces[0].IPs...)
+		curIPs := sets.NewString(curEn.Spec.Interfaces[0].IPs...)
+		if (!reflect.DeepEqual(preEn.Labels, curEn.Labels)) || (!preIPs.Equal(curIPs)) {
+			eeName := genExternalEntityName(curEn)
+			updatedEE := genExternalEntity(eeName, curEn)
+			err := c.updateExternalEntity(updatedEE)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	c.syncedExternalNode.Update(curEn)
+	return nil
+}
+func (c *ExternalNodeController) updateExternalEntity(ee *v1alpha2.ExternalEntity) error {
+	// resourceVersion must be specified for update operation,
+	// so it gets the existing ExternalEntity and modifies the changed fields.
+	existingEE, _ := c.crdClient.CrdV1alpha2().ExternalEntities(ee.Namespace).Get(context.TODO(), ee.Name, metav1.GetOptions{})
+	isChanged := false
+	if !reflect.DeepEqual(existingEE.Spec, ee.Spec) {
+		existingEE.Spec = ee.Spec
+		isChanged = true
+	}
+	if !reflect.DeepEqual(existingEE.Labels, ee.Labels) {
+		existingEE.Labels = ee.Labels
+		isChanged = true
+	}
+	if isChanged {
+		_, err := c.crdClient.CrdV1alpha2().ExternalEntities(ee.Namespace).Update(context.TODO(), existingEE, metav1.UpdateOptions{})
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (c *ExternalNodeController) deleteExternalNode(namespace string, name string) error {
+	obj, exists, _ := c.syncedExternalNode.GetByKey(k8s.NamespacedName(namespace, name))
+	if !exists {
```

Comment:
Thanks for the comment. I think this may happen when Add/Update and delete requests for the same ExternalNode are processed by two workers at the same time. I am thinking to change the worker number to one to avoid the potential concurrent issue, what do you think?